lib/scholar/model_selection.ex

```diff
@@ -110,4 +112,169 @@ defmodule Scholar.ModelSelection do
+      ...>   {x_train, x_test} = x
+      ...>   {y_train, y_test} = y
+      ...>   {weights_train, _weights_test} = weights
+      ...>   opts = Keyword.put(opts, :sample_weights, weights_train)
+      ...>   model = Scholar.Linear.LinearRegression.fit(x_train, y_train, opts)
+      ...>   y_pred = Scholar.Linear.LinearRegression.predict(model, x_test)
+      ...>   mse = Scholar.Metrics.Regression.mean_square_error(y_test, y_pred)
+      ...>   mae = Scholar.Metrics.Regression.mean_absolute_error(y_test, y_pred)
+      ...>   [mse, mae]
+      ...> end
+      iex> x = Nx.iota({7, 2})
+      iex> y = Nx.tensor([0, 1, 2, 0, 1, 1, 0])
+      iex> weights = Nx.tensor([1, 2, 1, 2, 1, 2, 1])
+      iex> Scholar.ModelSelection.weighted_cross_validate(x, y, weights, folding_fun, scoring_fun)
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [0.5010331869125366, 1.1419668197631836, 0.35123956203460693],
+          [0.5227273106575012, 1.0526316165924072, 0.5909090042114258]
+        ]
+      >
+
+  """
+  def weighted_cross_validate(x, y, weights, folding_fun, scoring_fun) do
+    Stream.zip([folding_fun.(x), folding_fun.(y), folding_fun.(weights)])
+    |> Enum.map(fn {x, y, weights} -> scoring_fun.(x, y, weights) |> Nx.stack() end)
+    |> Nx.stack(axis: 1)
+  end
+
+  defp combinations([]), do: [[]]
+
+  defp combinations([{name, values} | opts]) do
+    for subcombination <- combinations(opts), value <- values do
+      [{name, value} | subcombination]
+    end
+  end
+
+  @doc """
+  General interface of grid search.
+  If you want to use `opts` in some functions inside `scoring_fun`, you need to pass it as a parameter
+  like in the example below.
+
+  ## Examples
+
+      iex> folding_fun = fn x -> Scholar.ModelSelection.k_fold_split(x, 3) end
+      iex> scoring_fun = fn x, y, opts ->
+      ...>   {x_train, x_test} = x
+      ...>   {y_train, y_test} = y
+      ...>   model = Scholar.Linear.LogisticRegression.fit(x_train, y_train, opts)
+      ...>   y_pred = Scholar.Linear.LogisticRegression.predict(model, x_test)
+      ...>   mse = Scholar.Metrics.Regression.mean_square_error(y_test, y_pred)
+      ...>   mae = Scholar.Metrics.Regression.mean_absolute_error(y_test, y_pred)
+      ...>   [mse, mae]
+      ...> end
+      iex> x = Nx.iota({7, 2})
+      iex> y = Nx.tensor([0, 1, 2, 0, 1, 1, 0])
+      iex> opts = [
+      ...>   num_classes: [3],
+      ...>   iterations: [10, 20, 50],
+      ...>   optimizer: [Polaris.Optimizers.adam(learning_rate: 0.005), Polaris.Optimizers.adam(learning_rate: 0.01)],
+      ...> ]
+      iex> Scholar.ModelSelection.grid_search(x, y, folding_fun, scoring_fun, opts)
+  """
+  def grid_search(x, y, folding_fun, scoring_fun, opts) do
+    params = combinations(opts)
+
+    for param <- params do
+      scoring_function = check_arity(scoring_fun, params)
+
+      [
+        hyperparameters: param,
+        score: Nx.mean(cross_validate(x, y, folding_fun, scoring_function), axes: [1])
+      ]
```

Comment:
```suggestion
      %{
        hyperparameters: param,
        score: Nx.mean(cross_validate(x, y, folding_fun, scoring_function), axes: [1])
      }
```

I prefer a map since those fields are fixed. :)